chalk-engine/src/logic.rs

```diff
@@ -303,6 +309,13 @@ impl<I: Interner> Forest<I> {
             }
 
             _ => {
+                let (infer, subst, InEnvironment { environment, goal }) =
+                    chalk_solve::infer::InferenceTable::from_canonical(
+                        context.program().interner(),
+                        goal.universes,
+                        &goal.canonical,
+                    );
+                let mut infer = TruncatingInferenceTable::new(context.max_size(), infer);
                 // `canonical_goal` is a goal. We can simplify it
```

Comment:
pre-existing, but "canonical goal is a goal" is a pretty mystifying sentence. I guess what I meant is "canonical goal is a meta-goal" or something like that. In other words, it's a goal whose handling is built-in to the solver, versus a domain goal that is defined by rules. I forget the term for this, I think metagoal is right but...